src/object.rs

```diff
@@ -2570,3 +2429,116 @@ impl<'a> BindingBuilder<'a> {
         }
     }
 }
+
+#[repr(C)]
+pub struct Class<T: ObjectType>(T::GlibClassType);
+
+impl<T: ObjectType> Class<T> {
+    /// Get the type id for this class.
+    pub fn get_type(&self) -> Type {
+        unsafe {
+            let klass = self as *const _ as *const gobject_sys::GTypeClass;
+            from_glib((*klass).g_type)
+        }
+    }
+
+    /// Casts this class to a reference to a parent type's class.
+    pub fn upcast_ref<U: ObjectType>(&self) -> &Class<U>
+    where
+        T: IsA<U>,
+    {
+        unsafe {
+            let klass = self as *const _ as *const Class<U>;
+            &*klass
+        }
+    }
+
+    /// Casts this class to a mutable reference to a parent type's class.
+    pub fn upcast_ref_mut<U: ObjectType>(&mut self) -> &mut Class<U>
+    where
+        T: IsA<U>,
+    {
+        unsafe {
+            let klass = self as *mut _ as *mut Class<U>;
+            &mut *klass
+        }
+    }
+
+    /// Casts this class to a reference to a child type's class or
+    /// fails if this class is not implementing the child class.
+    pub fn downcast_ref<U: ObjectType>(&self) -> Option<&Class<U>>
+    where
+        U: IsA<T>,
+    {
+        if !self.get_type().is_a(&U::static_type()) {
+            return None;
+        }
+
+        unsafe {
+            let klass = self as *const _ as *const Class<U>;
+            Some(&*klass)
+        }
+    }
+
+    /// Casts this class to a mutable reference to a child type's class or
+    /// fails if this class is not implementing the child class.
+    pub fn downcast_ref_mut<U: ObjectType>(&mut self) -> Option<&mut Class<U>>
+    where
+        U: IsA<T>,
+    {
+        if !self.get_type().is_a(&U::static_type()) {
+            return None;
+        }
+
+        unsafe {
+            let klass = self as *mut _ as *mut Class<U>;
+            Some(&mut *klass)
+        }
+    }
+
+    /// Gets the class struct corresponding to `type_`.
+    ///
+    /// This will return `None` if `type_` is not a subclass of `Self`.
+    pub fn from_type(type_: Type) -> Option<ClassRef<T>> {
+        if !type_.is_a(&T::static_type()) {
+            return None;
+        }
+
+        unsafe {
+            let ptr = gobject_sys::g_type_class_ref(type_.to_glib());
+            if ptr.is_null() {
+                None
+            } else {
+                Some(ClassRef(ptr::NonNull::new_unchecked(ptr as *mut Self)))
+            }
+        }
+    }
+}
+
+unsafe impl<T: ObjectType> Send for Class<T> { }
+unsafe impl<T: ObjectType> Sync for Class<T> { }
+
+// This should require Self: IsA<Self::Super>, but that seems to cause a cycle error
+pub unsafe trait SuperClassIs: ObjectType {
```

Comment:
```suggestion
pub unsafe trait ParentClassIs: ObjectType {
```

Parent is the terminology we use everywhere else

Reply:
Everywhere except the `$super_name` of the macro, which I was naming it after. :laughing: 

But yeah, that's probably a better term given it refers to the direct parent, not any ancestor type.